core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1767,7 +1767,8 @@ public Void onRemoteTaskCompleted(RemoteTaskCompletedEvent event)
             TaskState taskState = taskStatus.getState();
             StageExecution stageExecution = getStageExecution(taskId.getStageId());
             if (taskState == TaskState.FINISHED) {
-                stageExecution.taskFinished(taskId, taskStatus);
+                Optional<List<PrioritizedScheduledTask>> failOverrideReplacementTasks = stageExecution.taskFinished(taskId, taskStatus);
+                failOverrideReplacementTasks.ifPresent(prioritizedScheduledTasks -> prioritizedScheduledTasks.forEach(schedulingQueue::addOrUpdate));
```

Comment:
I'm not sure I understand.  The task will only be "failed" if it's already finished.  Why schedule a replacement?  Is it because failing a task cancels downstream tasks automatically?  Then aren't those downstream tasks what needs rescheduling, not the current task?

Reply:
If we want query to finish we need to either:
* explicitly mark it as failed after task failure
* create new task which will do the work the failed task failed to complete

Reply:
Has this task failed to complete, though?
https://github.com/trinodb/trino/blob/f8d373f74b643be880d9f8c395dafb1c380e3e88/core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java#L2332-L2333
If this is true, then we're choosing neither of the two options above.  It looks like we're creating a new task to redo the successfully completed work.  I'm sure I'm missing something, but I doubt it's something obvious.

Reply:
Yeah - we are creating new task in place of the one which failed. It is new "attempt" but technically it is separate task.